Allow more fine-grained control over what submodules are updated: part 2

tjfulle opened the pull request:
@sethrj, this is an alternative to #23873.  I actually like this one better.  It allows the package to define a version's submodules as a callable function which is then called with the package as the first argument.  It allows for the `FetchStrategy` and `FetchStrategyComposite` to define a `set_package` attribute.  This should be compatible with source, resources, etc.  Usage:

```python
def submodules(package):
    if "+baz" in package.spec:
        return ["baz"]
    
class MyPackage(Package):
    variant("baz", default=False)
    version("1.0.0", git=git_url, submodules=submodules)
```

There are still some encapsulation issues, but I think this is a necessary evil since fetchers are instantiated before a package is concretized.

A disadvantage of this method compared to #23873 is that the submodule function will need to be passed to each `version` declaration - which is not different than the current requirement to set `submodules=True` in each `version` declaration.  A clear advantage is that when the package is assigned to its `fetcher` there need not be a distinction between the `root_fetcher`, `resources`, etc. (see comment https://github.com/spack/spack/pull/23873#discussion_r744862777)

tjfulle committed:
Adding Package.submodules command for more fine-grain control of submodule update

tjfulle committed:
init, then update, submodules.

tjfulle committed:
adding documentation or Package.submodules

tjfulle committed:
update package to have get_submodules method that checks user defined submodules()

tjfulle committed:
wrapping rst docs

tjfulle committed:
callable submodules

tjfulle committed:
update docs

tjfulle committed:
remove get_submodules

tjfulle committed:
submodules free function should be able to reutnr None.

tjfulle committed:
remove unused import

tjfulle committed:
flake8 fixes

tjfulle committed:
fix import ordering for spack style test

tjfulle committed:
adding test for callable submodules

tjfulle commented:
@alalazo or @sethrj - would one of you mind looking at this PR?

alalazo commented:
Hi @tjfulle I don't think I'll have much time before Spack's BoF at SC21 next Wednesday. If I forget about this PR feel free to ping me again after the BoF.

sethrj commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -1007,9 +1017,22 @@ def clone(self, dest=None, commit=None, branch=None, tag=None, bare=False):
-                args = ['submodule', 'update', '--init', '--recursive']
+                args = ['submodule', 'init']
+                if isinstance(submodules, list):
+                    args += ["--"] + list(submodules)
```
the `list` here is redundant since it's already a list and you're only using it as a temporary.

sethrj reviewed (dismissed):
Looks great to me! Might be a *little* nicer if the example correctly uses a `@classmethod`/`@staticmethod` but I'm not 100% sure it can be (haven't written enough python lately...)

sethrj commented:
Sorry @tjfulle it's been a very busy week for me too. I'm going to close and reopen this just to give the CI another try.

tjfulle commented:
Thanks @sethrj - yes, the `list` is redundant, a vestige of a previous veraion.

RE `staticmethod/classmethod` - a `classmethod` wouldn't make sense.  a `staticmethod` could work, I suppose, but it would need to be defined before the version declaration.  Free functions work fine and are what I've now adopted

sethrj commented:
Sounds good. Unfortunately I can't merge because something's up with the bootstrapping that's blocking the CI -- I'll ask the other maintainers about it.

tjfulle committed:
address PR comments and clean up documentation

sethrj commented:
@alalazo now that SC is over, could you help @tjfulle figure out why the bootstrap tests are failing with this change set? Thanks!

tjfulle committed:
adding more test coverage

tjfulle commented:
Bump @alazo

tjfulle commented:
Ping @alalazo

tjfulle commented:
@alalazo - has there been any movement on this?

alalazo commented:
> has there been any movement on this?

@tjfulle Not really, sorry. Right now we are kind of busy with a few big PR that need to get in (#28504 #28673 #28941) Is this a blocker for you in any way?

tjfulle commented:
@alalazo - we are not "blocked".  My code group at SNL uses a fork of spack with this change applied.  We would like to ditch the fork and use mainline Spack and this is stopping the transition.  We had hoped that this would be merged before 0.18.0 because some of our air gapped sites have trouble getting Clingo working.

alalazo commented:
> We had hoped that this would be merged before 0.18.0 because some of our air gapped sites have trouble getting Clingo working.

I'm not sure I understand this comment. Would you like to see this change in a v0.17.x point release? Or before v0.18.0 means you'd like to see it in v0.18.0?  Also, feel free to pester me in say in a couple of weeks (I assume we'd be finished with the 3 PRs above by then).

tjfulle commented:
Hey @alalazo - sorry for not being more clear.  When this is merged we will begin to pull directly from mainline Spack.  If there is a point release, great.  If not, we'll just be sure to use a commit that includes the PR.  If the PR is merged *before* the original concretizer is removed that would be very helpful.

alazo commented:
HelloI am alzo not alalazo!Becs regardsAlzoLe 8 mars 2022 18:58, Tim Fuller ***@***.***> a écrit :
Hey @alalazo - sorry for not being more clear.  When this is merged we will begin to pull directly from mainline Spack.  If there is a point release, great.  If not, we'll just be sure to use a commit that includes the PR.  If the PR is merged before the original concretizer is removed that would be very helpful.

—Reply to this email directly, view it on GitHub, or unsubscribe.Triage notifications on the go with GitHub Mobile for iOS or Android.
You are receiving this because you were mentioned.Message ID: ***@***.***>

psakievich commented:
@alalazo @haampie can one of you re-run the tests on this and look into getting this merged?   @tjfulle if you want you can close and re-open and it will re-run the tests.

alalazo commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1070,13 +1070,32 @@ Commits
+
+  .. code-block:: python
+
+     def submodules(package):
+         submodules = []
+         if "+variant-1" in package.spec:
+             submodules.append("submodule_for_variant_1")
+         if "+variant-2" in package.spec:
+             submodules.append("submodule_for_variant_2")
+         return submodules
```
Why are we passing a package and not just the concrete spec?

tjfulle replied:
Only for the general case where the submodules may depend on something other than the concrete spec.

alalazo commented on lib/spack/docs/packaging_guide.rst:
```diff
@@ -1070,13 +1070,32 @@ Commits
 
      version('1.0.1', tag='v1.0.1', submodules=True)
 
+  If a package has per-variant submodules, define ``submodules`` to be a
```
Maybe we can slightly modify the docs to be more general:
```suggestion
  If a package needs more fine-grained control over submodules, define ``submodules`` to be a
```
The idea is that we can construct the list of submodules based on any attribute of the spec, not just variants.

alalazo commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -1007,9 +1017,22 @@ def clone(self, dest=None, commit=None, branch=None, tag=None, bare=False):
+        if callable(submodules):
+            submodules = submodules(self.package)
+            if isinstance(submodules, six.string_types):
+                submodules = shlex.split(submodules)
```
This case is not in the docs if I read correctly. Can we get rid of it and assume we'll always get a list from the callable?

alalazo commented on lib/spack/spack/fetch_strategy.py:
```diff
@@ -1007,9 +1017,22 @@ def clone(self, dest=None, commit=None, branch=None, tag=None, bare=False):
             with working_dir(dest):
-                args = ['submodule', 'update', '--init', '--recursive']
+                args = ['submodule', 'init']
+                if isinstance(submodules, list):
```
I would avoid lot of conditionals based on types. What do you think if we can restructure these lines to construct the git commands first, and finally invoke them? A simplifies sketch of the idea is:
```python
git_commands = []
if callable(submodules):
    ...
    git_commands.append(['submodule', 'init', '--'] +submodules)
    git_commands.append(['submodule', 'update', 'recursive'])
elif submodules is True:
    git_commands.append(['submodule', 'update', '--init', '--recursive'])

if not git_commands:
    return

with working_dir(dest):
    for args in git_commands:
        if not spack.config.get('config:debug'):
            args.insert(1, '--quiet')
        git(*args)
```

alalazo commented on lib/spack/spack/test/git_fetch.py:
```diff
@@ -251,6 +251,41 @@ def test_gitsubmodule(submodules, mock_git_repository, config,
+    """
+    def submodules_callback(package):
+        name = 'third_party/submodule0'
+        return [name] if submodules is list else name if submodules is str else None
```
As commented above, I would avoid the string and the `None` cases to keep the implementation simpler. Also, not sure how easy it is, but it would be better for readability if all the setup is in one or more fixtures, so it is clear what the test is asserting.

alalazo reviewed (changes requested):
I left a few requests for minor changes. Let me know what you think. Apologies for the delay in reviewing this.

tjfulle committed:
merging master and addressing some comments by @alalazo

tjfulle committed:
fixing broken file from merge

tjfulle commented:
@alalazo  - I addressed the comments but ran into a snag merging the latest `spack:develop` with my fork getting out of date with `spack:develop`  -  so I force pushed to my branch.  This caused your review to get lost with the lost commits.

alalazo commented:
Thanks @tjfulle

alazo commented:
Hello

I am alazo, not alalazo!

Best regards

Alazo

Le 2022-04-08T06:59:22.000+02:00, psakievich
***@***.***> a écrit :

***@***.******@***.***
> [https://github.com/haampie] can one of you re-run the tests on this
***@***.***
> [https://github.com/tjfulle] if you want you can close and re-open
> and it will re-run the tests.
> 
> —
> Reply to this email directly, view it on GitHub
> [https://github.com/spack/spack/pull/27293#issuecomment-1092442568],
> or unsubscribe
> [https://github.com/notifications/unsubscribe-auth/AA6ROHHI25ZBH2E6ZRZRE5DVD64KVANCNFSM5HTEA34A].
> You are receiving this because you were mentioned.Message ID:
***@***.***>
